### PR TITLE
add a random sleep so that multiple threads don't work at the same time

### PR DESCRIPTION
**Pull request by StevenMassaro**
First off - thanks for making this tool. I'm pretty excited about it.

I have a few accounts set up where I'm running this, and I've noticed that occasionally one of the accounts will get a failure on startup. Unfortunately I don't have the logs because I've been testing this and the logs have rolled over. Sorry about that.

I suspected that this was occurring because each thread of `reservation_monitor` was running at the same time, which Southwest would likely be suspicious of. I added this random sleep to the `reservation_monitor` logic to try to make this more "human-like".

A couple of notes/concerns here:

- I did not discuss this with you before implementing this change, so consider this my "discussion". Though I am running this code myself and it appears to be working nicely.
- I don't think this will impact the actual check-in logic processing at the right time, but I would like your confirmation on that.
- It has been a long time since I've written any python, so any feedback on that front would be welcome (or any feedback at all).
- Maybe this should be configurable using a setting in the json config file? I can't come up with a good reason you'd want to turn this off but nevertheless.
- This sleep is not really needed if there is only one thread running, but I'm not sure the best way to determine that.

### CODE REVIEW

**Comment by jdholtz** [comment]
Thanks for the PR! Before accepting this, I would like to know what the error you are getting is because I can run multiple accounts without any issues. Additionally, this wouldn't be the only place in the script that processes could start at the same time (e.g. checking into the same flight for two separate reservations). 

> This sleep is not really needed if there is only one thread running, but I'm not sure the best way to determine that.

One way to do this is to read the process name and determine the process number from that (Process-1 -> 1, Process-2 -> 2, etc.). This is done in the `_wait_to_stop` method in the `ReservationMonitor` class. It probably isn't the best solution, but it would be better than picking a random sleep because: 
- There would be no delay for the first thread (solves the issue you mentioned)
- Ensures that two processes do not start at the same time

That being said, I would like to have this script work multiple processes running at the same time. Otherwise, I don't really see a point in having multiple processes for the reservation/account monitor. I doubt Southwest would be blocking it explicitly because they have thousands of users a second and one user connecting to their site three times from the same device should not be out of the ordinary.

**Comment by StevenMassaro** [comment]
Here's the error, I managed to get it to appear after restarting the docker container a few times (emulating the situation where I originally noticed this, when my server automatically goes down once a week to do a backup); attached are the full logs. Note that in the logs I pasted below, completely failing for a particular account was avoided by the retry logic. I didn't keep restarting the docker container until I was able to time it perfectly so that it failed 2 more times, but I have seen this happen before where both threads continue to retry at the same exact time until they both run out of retries.

There should probably also be a random sleep included in the retry logic so that both threads don't retry at the same time. I can add this if you agree.

```
...
2023-08-14 02:21:15 DEBUG MainProcess[main:142]: Monitoring 4 accounts and 0 reservations
2023-08-14 02:21:19 DEBUG Process-1[reservation_monitor:176]: Retrieving reservations for account
2023-08-14 02:21:19 DEBUG Process-1[webdriver:138]: Starting webdriver for current session
2023-08-14 02:21:19 DEBUG Process-2[reservation_monitor:176]: Retrieving reservations for account
2023-08-14 02:21:19 DEBUG Process-2[webdriver:138]: Starting webdriver for current session
2023-08-14 02:21:19 DEBUG Process-3[reservation_monitor:176]: Retrieving reservations for account
2023-08-14 02:21:19 DEBUG Process-3[webdriver:138]: Starting webdriver for current session
2023-08-14 02:21:19 DEBUG Process-4[reservation_monitor:176]: Retrieving reservations for account
2023-08-14 02:21:19 DEBUG Process-4[webdriver:138]: Starting webdriver for current session
2023-08-14 02:21:19 DEBUG Process-2[webdriver:172]: An exception occurred when initializing the webdriver: Name: OSError. Error: [Errno 26] Text file busy: '/usr/bin/chromedriver'
2023-08-14 02:21:19 DEBUG Process-2[webdriver:178]: 2 more attempts
2023-08-14 02:21:19 DEBUG Process-4[webdriver:172]: An exception occurred when initializing the webdriver: Name: OSError. Error: [Errno 26] Text file busy: '/usr/bin/chromedriver'
2023-08-14 02:21:19 DEBUG Process-4[webdriver:178]: 2 more attempts
2023-08-14 02:21:25 DEBUG Process-1[webdriver:147]: Loading Southwest Check-In page
2023-08-14 02:21:25 DEBUG Process-4[webdriver:147]: Loading Southwest Check-In page
2023-08-14 02:21:25 DEBUG Process-3[webdriver:147]: Loading Southwest Check-In page
2023-08-14 02:21:25 DEBUG Process-2[webdriver:147]: Loading Southwest Check-In page
/usr/local/lib/python3.11/site-packages/seleniumwire/thirdparty/mitmproxy/server/server.py:142: RuntimeWarning: coroutine 'AddonManager.handle_lifecycle' was never awaited
  self.log(traceback.format_exc(), "error")
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
2023-08-14 02:21:44 DEBUG Process-2[webdriver:90]: Logging into account to get a list of reservations and valid headers
...
```

> Additionally, this wouldn't be the only place in the script that processes could start at the same time (e.g. checking into the same flight for two separate reservations).

I would think this is a place where you _don't_ want to sleep, because you want to check in exactly at the right time. Not at the check-in time plus forty seconds, for example. I could be misunderstanding this situation though.

> I doubt Southwest would be blocking it explicitly because they have thousands of users a second and one user connecting to their site a few times from the same device should not be out of the ordinary.

Most websites _actively_ try to prevent this kind of access using off the shelf tools like [Cloudflare's bot management](https://www.cloudflare.com/application-services/products/bot-management/). It would be smart to skirt under the radar, and the best way to do that is to do things that normal humans do, and normal humans do not log in to four Southwest accounts simultaneously from the same exact device and IP address, hence my motivation for this change. My thinking is/was: put random sleeps in before doing things that don't need to occur at precise times (everything except the check in process).

Let me know what you're thinking. Once we're on the same page here, and if you're thinking you want to accept the change, I will go ahead and make the change wherever else it is necessary (though I'm not sure where else it is necessary...) and improve the logic a bit so that we don't needlessly sleep if only using one account.

[full logs.txt](https://github.com/jdholtz/auto-southwest-check-in/files/12331106/full.logs.txt)

**Comment by StevenMassaro** [comment]
Not sure if this matters, but I'm running the latest docker image on linux.

**Comment by StevenMassaro** [comment]
Also, here's a different error I've seen before. Though, it's unclear whether my fix will have any impact on this error:

```
2023-08-14 02:39:30 DEBUG Process-4[webdriver:90]: Logging into account to get a list of reservations and valid headers
2023-08-14 02:39:30 DEBUG Process-3[webdriver:90]: Logging into account to get a list of reservations and valid headers
2023-08-14 02:39:31 DEBUG Process-2[webdriver:90]: Logging into account to get a list of reservations and valid headers
2023-08-14 02:39:31 DEBUG Process-1[webdriver:90]: Logging into account to get a list of reservations and valid headers
2023-08-14 02:40:04 DEBUG Process-4[webdriver:225]: Logging in failed for an unknown reason
2023-08-14 02:40:04 WARNING Process-4[reservation_monitor:185]: Encountered a Too Many Requests error while logging in. Skipping reservation retrieval
2023-08-14 02:40:04 DEBUG Process-4[reservation_monitor:106]: Sleeping for 86314 seconds
2023-08-14 02:40:04 DEBUG Process-3[webdriver:225]: Logging in failed for an unknown reason
2023-08-14 02:40:04 WARNING Process-3[reservation_monitor:185]: Encountered a Too Many Requests error while logging in. Skipping reservation retrieval
2023-08-14 02:40:04 DEBUG Process-3[reservation_monitor:106]: Sleeping for 86313 seconds
```

**Comment by jdholtz** [comment]
> Here's the error

I see multiple errors in the traceback. The first one (Text file busy) should have been fixed in an earlier version of `undetected_chromedriver`, so maybe there were still some issues in `3.1.7` (the version this script uses). For the second error, that did occur in #89, which is why I added the retries to initializing the webdriver (it appears to happen after though). However, it isn't clear if any of the 4 processes failed from that log output. Did they still go on to work normally?

> I would think this is a place where you don't want to sleep, because you want to check in exactly at the right time

I agree. My main point in stating that was to say that if there is a problem with hitting Southwest's API concurrently with the same device and IP, it might be a problem in more areas than just logging in. 

> My thinking is/was: put random sleeps in before doing things that don't need to occur at precise times

I wouldn't really see an advantage to doing this vs. just running all reservation monitors in the same process. It might help to save a little bit of time running concurrently, but doing the latter would ensure no processes execute at the same time and the first account would always be executed right away (there could even be a tiny random delay added before starting each monitor to account for the 'humanness'.

> Also, here's a different error I've seen before, and was just able to reproduce. Though, it's unclear whether my fix will have any impact on this error:

This is a problem that people have been running into more lately. I wonder if upgrading to the latest `undetected_chromedriver` would fix it (AFAIK it has better bot detection algorithms and more reliability). The latest version of selenium-wire isn't compatible with it but I do have a fix that will hopefully work. Then we can test these issues again to make sure this wasn't an issue already fixed by just updating `undetected_chromedriver`.

I'm not against this change at all, I just want to make sure this is the correct fix for the issue. At the moment, it seems like a band-aid fix just to make the error less likely to occur (not 100% though due to randomness). If we can dig deeper and fix it without needing to rely on randomness, that would be great.

**Comment by StevenMassaro** [comment]
This is a band-aid, and you're right, running all the reservation monitors in the same process, not concurrently, would really solve it. There isn't a whole lot of benefit to running them concurrently.

Do you want me to take a stab at this change?

**Comment by jdholtz** [comment]
I would like to see if the issue you were running into that prompted you to make this change could be fixed first. What exactly was wrong in the first traceback? All processes seemed to start up even though a couple had to retry initialization.

**Comment by StevenMassaro** [comment]
If you're referring to this error:

```
2023-08-14 02:21:19 DEBUG Process-2[webdriver:178]: 2 more attempts
2023-08-14 02:21:19 DEBUG Process-4[webdriver:172]: An exception occurred when initializing the webdriver: Name: OSError. Error: [Errno 26] Text file busy: '/usr/bin/chromedriver'
```

Yes, after retrying initialization, startup proceeded normally in that instance. I didn't keep restarting the docker container to try to reproduce the issue I've seen before, where, that error shows up and both threads continue to retry, see that error, and fail to start.

**Comment by jdholtz** [comment]
What we can possibly do is isolate OSErrors to then retry at a random time between 1-5 seconds. It wouldn’t really solve the issue of starting at different times to be more under the radar, but the Too Many Requests error might not be fixed from this PR, as you stated above, but it seems like more testing would need to be done to see if it still happens. 

I would like to see if I can reproduce the issue from a Docker container (it’s strange no one else has reported the issue). Also, I think better multiprocessing support has been added in recent versions of undetected_chromedriver, so we can test those to see if the text file issue goes away (a few changes needed to be made in this script for compatibility with the new versions).

**Comment by StevenMassaro** [comment]
Another nice simple solution would be to only initialize one thread at a time. [Something like this](https://theorangeduck.com/page/synchronized-python):

```
with self.lock:
    # Do initialization
```

here: https://github.com/jdholtz/auto-southwest-check-in/blob/master/lib/webdriver.py#L151

I am looking at this from how I'd change this in Java, but from my limited googling that appears to be the equivalent in Python. This same thing could be used in the `reservation_monitor` too in place of the random sleeping, without requiring a whole big rearchitecting of the current threading model.

**Comment by StevenMassaro** [comment]
FYI - I know you're not sold on the sleeping as a potential solution here, but I moved the sleeps into a more logical location on my branch. I do think this brings it closer to "human like".

**Comment by jdholtz** [comment]
I just played around with this a bit and was able to reproduce the `Text file busy` issue very easily by running 10 instances at once (I did sys.exit() after the `_init_driver` call so it didn't actually reach out to Southwest). Surprisingly, the new versions of undetected_chromedriver actually make it function worse (there are more bugs), so upgrading is not the way to go.

I like the new location of the sleep. That works well. However, I just did a `time.sleep(0.1)` at the places you called `_sleep` and that worked as well with 10+ instances running concurrently. 

I guess the only issue now is deciding whether or not the large sleep delay is needed. The random sleep does make it more human-like, but I'm not sure it is necessary. What could be done is remove the `_safe_sleep` function entirely (it is unnecessary) and just sleep for the full retrieval interval without taking the time taken into account.

This would basically add a difference in execution time for every process because the time taken for every process will be different (needs to log in, schedule flights, check fares, etc. which will not take the exact same time for every process). Therefore, they won't trigger at the same time after the initial startup. 

The only issue could be the initial startup but the 429 error happens very rarely and can easily be worked around by just restarting the script. But once it is running, the processes shouldn't start at the exact same time again and will therefore not have the same issue as the initial startup. I don't think it's worth it to have extra sleep time at the beginning because that delays the user from seeing the script's results and fixes an issue that occurs very rarely (with an easy workaround).

Let me know your thoughts on this and if I should clarify anything. Thanks for working with me on this.

**Comment by StevenMassaro** [comment]
> I just played around with this a bit and was able to reproduce the Text file busy issue very easily by running 10 instances at once (I did sys.exit() after the _init_driver call so it didn't actually reach out to Southwest). Surprisingly, the new versions of undetected_chromedriver actually make it function worse (there are more bugs), so upgrading is not the way to go.

On one hand, I'm happy you were able to reproduce it. On the other hand, boo.

> I like the new location of the sleep. That works well. However, I just did a time.sleep(0.1) at the places you called _sleep and that worked as well with 10+ instances running concurrently.
> 
> I guess the only issue now is deciding whether or not the large sleep delay is needed. The random sleep does make it more human-like, but I'm not sure it is necessary.

It definitely does not need to be as long as I made it; 60 seconds would probably be sufficient. However, I still do think that human-ness is something to strive for with this script so it doesn't get shut down. It's not hard to detect that Selenium is being used on a site, and if Southwest gets their act together they could shut this down very easily. However, read on to the next point for a different approach...

> What could be done is remove the _safe_sleep function entirely (it is unnecessary) and just sleep for the full retrieval interval without taking the time taken into account.
> 
> This would basically add a difference in execution time for every process because the time taken for every process will be different (needs to log in, schedule flights, check fares, etc. which will not take the exact same time for every process). Therefore, they won't trigger at the same time after the initial startup.

My recommendation here would be:
- do the startup checks, one account at a time, not concurrently
- sleep for the retrieval interval
- again, do the check one account at a time

The concurrency of checking an account for its reservations seems unnecessary to me, particularly since it'll be sleeping for the better part of a day. If it takes 5 minutes to start up instead of 30 seconds, makes no difference to me since it's running constantly anyway. The concurrency in the check-in process is super necessary though!

In my experience, there are two things in software development that are "I had one problem and I solved it with X and now I have two". Regex and concurrency, so if you can solve the problem you have without using those tools, that tends to be better in the long run.

However, my python is rusty enough, and I am unfamiliar with this script that I am not immediately sure how I'd go about refactoring this to only execute one at a time, rather than concurrently.

> The only issue could be the initial startup but the 429 error happens very rarely and can easily be worked around by just restarting the script. But once it is running, the processes shouldn't start at the exact same time again and will therefore not have the same issue as the initial startup. I don't think it's worth it to have extra sleep time at the beginning because that delays the user from seeing the script's results and fixes an issue that occurs very rarely (with an easy workaround).

"can easily be worked around by just restarting the script" was actually my main motivation for fixing this. Every time I have to SSH into my server (after a bit of groaning), my mind always goes to "how can I fix whatever this was so I don't have to do this again". So to me, I would rather have a slower startup process (5 to 10 minutes) than having to restart the script every once in awhile.

But even ignoring the inconvenience of sshing (which is admittedly small), my main concern is that I have to realize that something went wrong. When the thread for one of the four accounts I have configured fails to start, I have to _notice_. It actually took me a bit to realize "hey, I should be getting four messages when this starts, and this week I only got three, something is up". To me, there are two solutions here. One, do the sleeping to try avoid the problem in the first place. Or two, (or maybe both), send a message via the notification system when a thread fails that says "hey! the application didn't start properly and a restart usually solves it".

Basically to sum up all of my opinions:
- I don't want to have to remember to check in on the script to make sure it started properly
- I do not want Southwest to ban this
- I think all of these things could be avoided by removing the concurrency in the startup process, and sleeps are my python-uneducated-way of doing this, though a refactor would probably be better.

**Comment by jdholtz** [comment]
I agree to make the account/reservation checking sequential instead of concurrent as I don't really see an advantage to it and it is better to avoid concurrency problems, as you mentioned.

Do you want me to work on this or do you want to implement this?

**Comment by StevenMassaro** [comment]
@jdholtz , take a look at the changes I just made. I think it strikes a nice balance of not requiring a large refactor, but also preventing each thread from working at the same time as any other thread. In my testing this effectively introduces concurrency for each reservation and account, but also keeps around the per reservation and per account retrieval interval.

It is totally fine with me if you had something else in mind here and would rather do it differently/yourself.

**Comment by jdholtz** [comment]
I like that change. Is it good if I format the code and fix the tests before merging it or was there anything else you wanted to add?

**Comment by StevenMassaro** [comment]
Sure! Change whatever you like - fine with me.

**Comment by jdholtz** [comment]
The lock also works well for stopping accounts/reservations in order when Ctrl+C is pressed, so I used the lock as a class variable and improved the stopping. This seems to be the ideal solution. Much appreciated for working through this with me!

**Comment by StevenMassaro** [comment]
You're welcome! Looking at the code a bit more, I think we'll need a lock here too: https://github.com/jdholtz/auto-southwest-check-in/blob/master/lib/checkin_scheduler.py#L49

Don't you agree? That's definitely a place where you don't want some kind of a selenium error. I think it should be safe to add one there because that call out occurs 10 minutes before the actual check in.

Maybe it would be better to move the `with lock:` part into the webdriver class, since that is ultimately where the locking needs to happen.

**Comment by jdholtz** [comment]
Good catch. The locking could only be done in the webdriver class. However, checking fares and scheduling check-ins is not done in the webdriver class and moving the lock into the webdriver class only would mean those tasks may run at the same time for multiple monitors. 

On the contrary, even the fare checking and scheduling sequential since it takes a long enough time for each monitor to use the webdriver that, by the time the next monitor has finished with the webdriver (and therefore the lock), the previous monitor would have finished.

The other way this can easily be done is just passing the lock into the Checkin Handler as well since the lock is used in other places besides the Webdriver class (such as in `ReservationMonitor:_stop_monitoring`)

**Comment by jdholtz** [comment]
@StevenMassaro do you have any comments on my above ideas?

**Comment by StevenMassaro** [comment]
> @StevenMassaro do you have any comments on my ideas above?

I had some trouble following your suggestions above. I think what I'd do is add a new lock here: https://github.com/jdholtz/auto-southwest-check-in/blob/develop/lib/checkin_handler.py#L90 and not try to reuse any of the existing lock instances that are used in the reservation monitor and account monitor classes.

What do you think?

**Comment by jdholtz** [comment]
> I had some trouble following your suggestions above.

To clarify my suggestions above, I am contemplating two choices:
1. Add the lock class to the line you pointed out in the CheckInHandler class
2. Add the lock in the WebDriver class (like you suggested)

The problem I can see with 2 is that the lock would only apply for the webdriver, but not during the requests for the fare checker and checkin scheduler. While there still wouldn't be a problem of two drivers starting at the same time, the problem of hitting the Southwest API concurrently would still be an issue.

> I think what I'd do is add a new lock here: https://github.com/jdholtz/auto-southwest-check-in/blob/develop/lib/checkin_handler.py#L90 and not try to reuse any of the existing lock instances that are used in the reservation monitor and account monitor classes.

Why would existing lock instances not be used? Wouldn't that mean that the webdrivers have a chance of starting at the same time? With the current setup in the script, I don't think it is feasible to have a lock instance for reservation monitors and a separate one for checkin handlers. However, I believe there *could* be an issue if we try to use separate lock instances for both based on my questions above.

**Comment by StevenMassaro** [comment]
You're right, a second lock might cause a deadlock (not totally sure on this, I didn't spend a ton of time thinking about it). Using the same lock class should be fine so long as you don't start the app right before check in time for one of your flights. I think that if you did that, the check in handler might be able to get a lock fast enough since it needs to wait for the monitor classes to finish starting up. This is a very unlikely scenario.

Sounds like your option 1 is a simple solution to the problem.

**Comment by jdholtz** [comment]
I agree. I'll even move the check-in handler to start 30 minutes before check-in to refresh its headers instead of 10 minutes just to ensure the scenario will most likely never happen. Thanks for your feedback and pointing this out.

**Comment by StevenMassaro** [comment]
> I agree. I'll even move the check-in handler to start 30 minutes before check-in to refresh its headers instead of 10 minutes just to ensure the scenario will most likely never happen. Thanks for your feedback and pointing this out.

Sure. Quick question - do you know how long the headers are good for? Is 30 minutes _too_ much? I have no idea, just want to make sure you thought of it since the check-in flow is like the most important part.

**Comment by jdholtz** [comment]
> do you know how long the headers are good for? Is 30 minutes too much?

They are valid for about 24-36 hours the last time I tested (about a year ago), so 30 minutes is perfectly fine

**Comment by jdholtz** [comment]
After switching to SeleniumBase, this change probably isn't necessary anymore. However, I will keep it because
1. Locking the threads is already in the code
2. Running browsers in parallel has more of a chance of causing issues than running them subsequently and the frequency at which they run at is low, so the tiny increase in the speed gained is not worth the potential for issues
3. It *could* be suspicious that multiple login attempts are happening within 2-3 seconds from the same IP, instead of the 15 seconds it takes when the browsers are run in succession.
